spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/SubsetServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2012-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.loadbalancer.core;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+
+import reactor.core.publisher.Flux;
+
+import org.springframework.cloud.client.ServiceInstance;
+
+/**
+ * A {@link ServiceInstanceListSupplier} implementation that uses
+ * <a href="https://sre.google/sre-book/load-balancing-datacenter/">deterministic
+ * subsetting</a> to limit the number of instances provided by delegate.
+ *
+ * @author Zhuozhi Ji
+ */
+public class SubsetServiceInstanceListSupplier extends DelegatingServiceInstanceListSupplier {
+
+	static final int DEFAULT_SUBSET_SIZE = 100;
+
+	final String instanceId;
```

Comment:
The naming of this field does not seem very clear to me; we use the term `instanceId` to signify the id of each of the instances available in the ServiceRegistry, for a given `serviceId`. I see here you use it to generate a `clientId`. I suggest you rename the field to `clientId` and perform any necessary transformations while setting its value.

Reply:
I'm thinking, we should also add a configuration properties  for the `clientId` and `subsetSize` along with their default values, that allow the users to set it and in this class we could just inject the corresponding `@ConfiguationProperties` bean through constructor instead of passing the values itself. We'd then just need one constructor (see `HintBasedServiceInstanceListSupplier` as a reference).

Reply:
Got, I will fix them

Reply:
Basically, there are two things. One is the id of the `LoadBalancer` context responsible for the communication with a given external service that encompasses all the `ServiceInstanceListSupplier` and `ReactiveLoadBalancer` instances used to communicate with a given service; so if the app uses load-balancing to communicate with `test-service` and `demo-service`, 2 such named LoadBalancer contexts will be created, however, unless set up otherwise it happens lazily during the first load-balanced call to that service. See [the code](https://github.com/spring-cloud/spring-cloud-commons/blob/14bfa57922f344f92b93aada588a4a7d0ebaf742/spring-cloud-context/src/main/java/org/springframework/cloud/context/named/NamedContextFactory.java#L235-L239). Another thing is how the client application registers itself in a service registry - it's general `clientId`, which to me is what we should use. That is simply taken from Spring's `spring.application.name` property. See the code: https://github.com/spring-cloud/spring-cloud-commons/blob/14bfa57922f344f92b93aada588a4a7d0ebaf742/spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryClientAutoConfiguration.java#L61-L69 and https://github.com/spring-cloud/spring-cloud-commons/blob/14bfa57922f344f92b93aada588a4a7d0ebaf742/spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/reactive/SimpleReactiveDiscoveryClientAutoConfiguration.java#L59-L60, so to me that's what you should take. Then, for every LoadBalancer context, you may create a `SubsettingServiceInstanceListSupplier` bean with the same `clientId` (taken from `spring.application.name`), but this will not be a problem, since each of these beans would communicate with a different service. Let me know if you have any questions or comments on this.

Reply:
And the property field for the `subsetSize` should go to `LoadBalancerProperties` and be retrieved from there.

Reply:
Hello @OlgaMaciaszek .`spring.application.name` cannot satisfy this scenario. In fact, there are now a large number of request instances (thousands or tens of thousands), each consumer instance needs to select a deterministic subset from them. At the same time, because the consumer may have tens of thousands of instances providing consistent services, it cannot be distinguished by `spring.application.name`.

So in my mind, I prefer to use the following information registered in Eureka. Is there any way to get an id like `192.168.0.1:eureka:8761`?

![image](https://github.com/spring-cloud/spring-cloud-commons/assets/31175030/373de82e-5daf-427c-8460-030d7c366a25)

Reply:
Ok, I see the issue, but I don't think getting this info from Eureka is the solution, since that's just one of the supported SR implementations, so we should only use whatever information we can get from `DiscoveryClient` that supports all the implementations, and we don't get this kind of data there; your initial `UUID.random()` default might actually be the best in this case; @spencergibb @ryanjbaxter wdyt?

Reply:
@jizhuozhi, I've chatted with @spencergibb, who's rightly pointed out that we could use [this method](https://github.com/spring-cloud/spring-cloud-commons/blob/41ae9a429761e7fbd465cae181be26897f7ea6a4/spring-cloud-commons/src/main/java/org/springframework/cloud/commons/util/IdUtils.java#L44-L61) for setting the `clientId`. Please take a look.

Reply:
I think it will be useful that using `DEFAULT_SERVICE_ID_STRING ` as default but user could custom the expression
 https://github.com/spring-cloud/spring-cloud-commons/blob/41ae9a429761e7fbd465cae181be26897f7ea6a4/spring-cloud-commons/src/main/java/org/springframework/cloud/commons/util/IdUtils.java#L30